pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentLoader.java

```diff
@@ -0,0 +1,212 @@
+      throws Exception {
+    IndexLoadingConfig defaultIndexLoadingConfig = new IndexLoadingConfig();
+    defaultIndexLoadingConfig.setReadMode(readMode);
+    return load(indexDir, defaultIndexLoadingConfig, null, false);
+  }
+
+  /**
+   * Loads the segment with empty schema but a specified IndexLoadingConfig.
+   * This method modifies the segment like to convert segment format, add or remove indices.
+   * Mostly used by UT cases to add some specific index for testing purpose.
+   */
+  public static ImmutableSegment load(File indexDir, IndexLoadingConfig indexLoadingConfig)
+      throws Exception {
+    return load(indexDir, indexLoadingConfig, null, true);
+  }
+
+  /**
+   * Loads the segment with specified schema and IndexLoadingConfig, usually from Zookeeper.
+   * This method modifies the segment like to convert segment format, add or remove indices.
+   * Mainly used during segment reloading.
+   */
+  public static ImmutableSegment load(File indexDir, IndexLoadingConfig indexLoadingConfig, @Nullable Schema schema)
+      throws Exception {
+    return load(indexDir, indexLoadingConfig, schema, true);
+  }
+
+  /**
+   * Loads the segment with specified schema and IndexLoadingConfig, and allows to control whether to
+   * modify the segment like to convert segment format, add or remove indices.
+   */
+  public static ImmutableSegment load(File indexDir, IndexLoadingConfig indexLoadingConfig, @Nullable Schema schema,
+      boolean shouldModifySegment)
+      throws Exception {
+    Preconditions.checkArgument(indexDir.isDirectory(), "Index directory: %s does not exist or is not a directory",
+        indexDir);
+
+    SegmentMetadataImpl localSegmentMetadata = new SegmentMetadataImpl(indexDir);
+    if (localSegmentMetadata.getTotalDocs() == 0) {
+      return new EmptyIndexSegment(localSegmentMetadata);
+    }
+
+    // This step will modify the segment data on disk.
+    if (shouldModifySegment) {
+      // Convert segment version as needed.
+      convertSegmentFormat(indexDir, indexLoadingConfig, localSegmentMetadata);
+      // Preprocess the segment on local using local SegmentDirectory.
+      preprocessSegment(indexDir, localSegmentMetadata.getName(), indexLoadingConfig, schema);
+    }
+
+    // Load the segment again using the configured segmentDirectoryLoader
+    PinotConfiguration segmentDirectoryConfigs = indexLoadingConfig.getSegmentDirectoryConfigs();
+    SegmentDirectoryLoaderContext segmentLoaderContext =
+        new SegmentDirectoryLoaderContext(indexLoadingConfig.getTableConfig(), indexLoadingConfig.getInstanceId(),
+            localSegmentMetadata.getName(), segmentDirectoryConfigs);
+
+    SegmentDirectoryLoader segmentDirectoryLoader =
+        SegmentDirectoryLoaderRegistry.getSegmentDirectoryLoader(indexLoadingConfig.getSegmentDirectoryLoader());
+    SegmentDirectory actualSegmentDirectory = segmentDirectoryLoader.load(indexDir.toURI(), segmentLoaderContext);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *actualSegmentDirectory*. Currently, there are execution paths that do not contain closure statements, for example, when SegmentDirectory.createReader throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).